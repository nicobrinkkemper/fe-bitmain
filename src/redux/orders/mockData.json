[
    {
        "id": 1,
        "type": "sell",
        "quantity": 15,
        "price": 113
    },
    {
        "id": 2,
        "type": "buy",
        "quantity": 8,
        "price": 70
    },
    {
        "id": 3,
        "type": "sell",
        "quantity": 7,
        "price": 86
    },
    {
        "id": 4,
        "type": "sell",
        "quantity": 2,
        "price": 222
    },
    {
        "id": 5,
        "type": "buy",
        "quantity": 8,
        "price": 65
    },
    {
        "id": 6,
        "type": "buy",
        "quantity": 3,
        "price": 78
    },
    {
        "id": 7,
        "type": "buy",
        "quantity": 9,
        "price": 61
    },
    {
        "id": 8,
        "type": "sell",
        "quantity": 6,
        "price": 88
    },
    {
        "id": 9,
        "type": "buy",
        "quantity": 9,
        "price": 105
    },
    {
        "id": 10,
        "type": "buy",
        "quantity": 14,
        "price": 79
    },
    {
        "id": 11,
        "type": "buy",
        "quantity": 10,
        "price": 56
    },
    {
        "id": 12,
        "type": "sell",
        "quantity": 14,
        "price": 203
    },
    {
        "id": 13,
        "type": "buy",
        "quantity": 11,
        "price": 107
    },
    {
        "id": 14,
        "type": "sell",
        "quantity": 12,
        "price": 202
    }
]