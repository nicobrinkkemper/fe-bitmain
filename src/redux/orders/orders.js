import { ADD_ORDER, TOGGLE_MATCH, MATCH_ORDER, LIST_ORDERS_RECEIVED, RESET_LIST_ORDERS_RECEIVED, RESET_LIST_ORDERS_REQUEST } from '../actionTypes';
import { ORDER_TYPES } from '../../constants';
import { createSelector } from 'reselect'
// SELECTORS
export const hasQuantity = ({ quantity }) => quantity > 0
export const tsDesc = ({ ts: a }, { ts: b }) => b - a
export const idAsc = ({ id: a }, { id: b }) => a - b
export const isBuy = ({ type }) => type === ORDER_TYPES.BUY
export const isSell = ({ type }) => type === ORDER_TYPES.SELL
export const isMatch = ({ type }) => type === ORDER_TYPES.MATCH
export const priceDesc = ({ price: a }, { price: b }) => a - b
export const priceAsc = ({ price: a }, { price: b }) => b - a
const counterDeal = order => isSell(order)
    ? counterOrder => isBuy(counterOrder) && hasQuantity(counterOrder) && counterOrder.price >= order.price
    : counterOrder => isSell(counterOrder) && hasQuantity(counterOrder) && counterOrder.price <= order.price

export const selectMatches = createSelector(
    orders => orders
        .filter(isMatch),
    orders => orders
        .sort(tsDesc)
        .slice(0, 30)
)
export const selectSells = createSelector(
    orders => orders
        .filter(isSell),
    orders => orders
        .filter(hasQuantity)
        .sort(idAsc)
        .sort(priceAsc)
        .slice(0, 20)
)
export const selectBuys = createSelector(
    orders => orders
        .filter(isBuy),
    orders => orders
        .filter(hasQuantity)
        .sort(idAsc)
        .sort(priceDesc)
        .slice(0, 20)
)

// ACTIONS
export const [addOrder, toggleMatch, matchOrder] = [
    ADD_ORDER, TOGGLE_MATCH, MATCH_ORDER
]
    .map(
        type => payload => (
            {
                type,
                payload
            }
        )
    )

// EPICS

// REDUCERS
export const orders = (state = [], action) => {
    switch (action.type) {
        case TOGGLE_MATCH:{
            const order = action.payload;
            return state.map(
                _order => _order.id === order.id
                    ? { ..._order, toggled: !_order.toggled }
                    : _order
            )
        }
        case RESET_LIST_ORDERS_REQUEST:
        case RESET_LIST_ORDERS_RECEIVED:
            return []
        case LIST_ORDERS_RECEIVED:
            return action.payload.response.map(addOrder).reduce(orders, state)
        case ADD_ORDER:{
            const order = action.payload;
            if(!hasQuantity(order)) return [...state, order]
            const counterOrder = state.find(counterDeal(order))
            if (!counterOrder) return [...state, order]
            return orders(state, matchOrder({order,counterOrder}))
        }
        case MATCH_ORDER:{
            const {order,counterOrder} = action.payload;
            const price = (order.price + counterOrder.price) / 2
            const quantity = Math.min(order.quantity, counterOrder.quantity)
            const ts = Date.now()
            const orderResult = { ...order, quantity: order.quantity - quantity }
            const counterOrderResult = { ...counterOrder, quantity: counterOrder.quantity - quantity }
            const match = {
                id: order.id + '-' + counterOrder.id,
                type: ORDER_TYPES.MATCH,
                orderResult,
                counterOrderResult,
                order,
                counterOrder,
                price,
                quantity,
                ts,
                toggled: false
            }
            const updatedState = [
                ...state.map(
                    _order => _order.id === counterOrder.id
                        ? counterOrderResult
                        : _order
                ),
                match
            ]
            return orders(updatedState, addOrder(orderResult))
        }
        default:
            return state
    }
}

export default orders