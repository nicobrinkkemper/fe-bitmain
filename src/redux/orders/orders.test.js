import orders, { addOrder } from './orders.js'
import MockDate from 'mockdate';
MockDate.set('1/1/2000');
const data = [
  {
    "id": 1,
    "type": "sell",
    "quantity": 6,
    "price": 480
  },
  {
    "id": 2,
    "type": "buy",
    "quantity": 8,
    "price": 470
  },
  {
    "id": 3,
    "type": "buy",
    "quantity": 5,
    "price": 460
  },
  {
    "id": 4,
    "type": "sell",
    "quantity": 10,
    "price": 460
  }
]

describe('reducers', () => {
  it('should match snapshot', () => {
    expect(
      data.map(addOrder).reduce(orders, [])
    ).toMatchSnapshot();
  })
})