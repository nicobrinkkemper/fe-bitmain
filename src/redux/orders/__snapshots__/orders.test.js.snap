// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers should match snapshot 1`] = `
Array [
  Object {
    "id": 1,
    "price": 480,
    "quantity": 6,
    "type": "sell",
  },
  Object {
    "id": 2,
    "price": 470,
    "quantity": 0,
    "type": "buy",
  },
  Object {
    "id": 3,
    "price": 460,
    "quantity": 3,
    "type": "buy",
  },
  Object {
    "counterOrder": Object {
      "id": 2,
      "price": 470,
      "quantity": 8,
      "type": "buy",
    },
    "counterOrderResult": Object {
      "id": 2,
      "price": 470,
      "quantity": 0,
      "type": "buy",
    },
    "id": "4-2",
    "order": Object {
      "id": 4,
      "price": 460,
      "quantity": 10,
      "type": "sell",
    },
    "orderResult": Object {
      "id": 4,
      "price": 460,
      "quantity": 2,
      "type": "sell",
    },
    "price": 465,
    "quantity": 8,
    "toggled": false,
    "ts": 946681200000,
    "type": "match",
  },
  Object {
    "counterOrder": Object {
      "id": 3,
      "price": 460,
      "quantity": 5,
      "type": "buy",
    },
    "counterOrderResult": Object {
      "id": 3,
      "price": 460,
      "quantity": 3,
      "type": "buy",
    },
    "id": "4-3",
    "order": Object {
      "id": 4,
      "price": 460,
      "quantity": 2,
      "type": "sell",
    },
    "orderResult": Object {
      "id": 4,
      "price": 460,
      "quantity": 0,
      "type": "sell",
    },
    "price": 460,
    "quantity": 2,
    "toggled": false,
    "ts": 946681200000,
    "type": "match",
  },
  Object {
    "id": 4,
    "price": 460,
    "quantity": 0,
    "type": "sell",
  },
]
`;
